### ISSUE TITLE
Optimize `gs_spending_bound()`

### ISSUE THREAD

**Issue by nanxstats** [Merck/gsDesign2#219]
In gsDesign, the C function `hupdate` was not directly called in R, but indirectly [called](https://github.com/keaven/gsDesign/blob/7ce330dc8cf3d42d66d3c5047649c480f5fd4487/R/gsDesign.R#L139) via a C function `gsbound`, which does the memory allocation and management effectively.

In gsDesign2, the corresponding part of the [logic](https://github.com/Merck/gsDesign2/blob/6358f06466bf8f395d8c5183879ce1688d31ef26/R/gs_spending_bound.R#L212) is currently implemented in R, which could be a bottleneck to slow things down, to the extent that a prospect web interface's responsiveness will be affected.

So, at some point, we will want to rewrite `gsbound` (and `gsbound1`) in C++.

**Comment by LittleBeannie** [comment]
Hi @nanxstats , is there any ideal person we can assign this ticket to? Let us do profiling first before the C++ rewrite.

**Comment by nanxstats** [comment]
@LittleBeannie No one in particular came to mind, do you have a suggestion?

**Comment by LittleBeannie** [comment]
I guess you are the only expert in this area in BAAMR 👍 

> @LittleBeannie No one in particular came to mind, do you have a suggestion?

**Comment by elong0527** [comment]
Per team discussion, @nanxstats will drive the implementation.

**Comment by LittleBeannie** [comment]
On hold for now

**Comment by jdblischak** [comment]
I did some benchmarking and profiling. I don't think the R code in `gs_spending_bound()` is a major bottleneck, at least in the examples I tried. It would be helpful for me to have some extra examples of code that would typically be executed via the web interface.

First I benchmarked and profiled the example in `?gs_spending_bound`, which calls `gs_power_ahr()`. There are many bottlenecks, but `gs_spending_bound()` and `hupdate()` barely register. The biggest cumulative bottlenecks are `expected_time()`, `expected_event()`, `gs_info_ahr()`, `pw_info()`, and `ahr()` (note that these are likely nested, so I'd need to dig deeper to find the best place to optimize first)

```R
devtools::load_all()
library("microbenchmark")
library("profvis")

microbenchmark(
  gs_power_ahr = gs_power_ahr(
    analysis_time = c(12, 24, 36),
    event = c(30, 40, 50),
    binding = TRUE,
    upper = gs_spending_bound,
    upar = list(sf = gsDesign::sfLDOF, total_spend = 0.025, param = NULL, timing = NULL),
    lower = gs_spending_bound,
    lpar = list(sf = gsDesign::sfLDOF, total_spend = 0.025, param = NULL, timing = NULL)
  ),
  times = 50
)
## Unit: milliseconds
## expr     min       lq     mean   median       uq      max neval
## gs_power_ahr 863.136 1004.379 1144.783 1104.985 1238.209 1835.487    50

profvis(
  gs_power_ahr(
    analysis_time = c(12, 24, 36),
    event = c(30, 40, 50),
    binding = TRUE,
    upper = gs_spending_bound,
    upar = list(sf = gsDesign::sfLDOF, total_spend = 0.025, param = NULL, timing = NULL),
    lower = gs_spending_bound,
    lpar = list(sf = gsDesign::sfLDOF, total_spend = 0.025, param = NULL, timing = NULL)
  )
)

profvis({
  for (i in 1:10)
  gs_power_ahr(
    analysis_time = c(12, 24, 36),
    event = c(30, 40, 50),
    binding = TRUE,
    upper = gs_spending_bound,
    upar = list(sf = gsDesign::sfLDOF, total_spend = 0.025, param = NULL, timing = NULL),
    lower = gs_spending_bound,
    lpar = list(sf = gsDesign::sfLDOF, total_spend = 0.025, param = NULL, timing = NULL)
  )
})
```

Second, I focused in on `gs_spending_bound()` by using the old examples that were removed in #261. This identified only two bottlenecks:

![image](https://github.com/Merck/gsDesign2/assets/1608317/c9e234f1-6132-402e-a104-eb829f88bd0a)

The biggest bottleneck was calling out to [`gsDesign::sfLDOF()`](https://github.com/keaven/gsDesign/blob/6de6583d621bf2fd0bff8e4b4f4b755d77ee407d/R/gsSpending.R#L701)

https://github.com/Merck/gsDesign2/blob/50067c424ab7dca4579ae895edc5de2bdcd9546e/R/gs_spending_bound.R#L134

The second bottleneck, which was only a fraction of the above, was calling out to `hupdate()`, which has already been converted to Rcpp

https://github.com/Merck/gsDesign2/blob/50067c424ab7dca4579ae895edc5de2bdcd9546e/R/RcppExports.R#L13


```R
old_examples <- function() {
  info <- (1:3) * 10
  info_frac <- info / max(info)
  k <- length(info_frac)
  
  # 1st analysis
  a1 <- gs_spending_bound(
    k = 1, efficacy = FALSE, theta = 0,
    par = list(sf = gsDesign::sfLDOF, total_spend = 0.025, timing = info_frac, param = NULL),
    hgm1 = NULL
  )
  b1 <- gs_spending_bound(
    k = 1, efficacy = TRUE, theta = 0,
    par = list(sf = gsDesign::sfLDOF, total_spend = 0.025, timing = info_frac, param = NULL),
    hgm1 = NULL
  )
  cat("The (lower, upper) boundary at the 1st analysis is (", a1, ", ", b1, ").\n")
  
  # 2nd analysis
  a2 <- gs_spending_bound(
    k = 2, efficacy = FALSE, theta = 0,
    par = list(sf = gsDesign::sfLDOF, total_spend = 0.025, timing = info_frac, param = NULL),
    hgm1 = h1(r = 18, theta = 0, info = info[1], a = a1, b = b1)
  )
  b2 <- gs_spending_bound(
    k = 2, efficacy = TRUE, theta = 0,
    par = list(sf = gsDesign::sfLDOF, total_spend = 0.025, timing = info_frac, param = NULL),
    hgm1 = h1(r = 18, theta = 0, info = info[1], a = a1, b = b1)
  )
  cat("The upper boundary at the 2nd analysis is (", a2, ", ", b2, ").\n")
}

profvis(for (i in 1:10) old_examples())
```

In summary, at least for the example call to `gs_power_ahr()`, converting more of `gs_spending_bound()` to C++ is unlikely to make a noticeable difference.

**Comment by keaven** [comment]
I think you’re figuring out where investment may be most beneficial KeavenOn Dec 13, 2023, at 13:01, John Blischak ***@***.***> wrote:﻿
I did some benchmarking and profiling. I don't think the R code in gs_spending_bound() is a major bottleneck, at least in the examples I tried. It would be helpful for me to have some extra examples of code that would typically be executed via the web interface.
First I benchmarked and profiled the example in ?gs_spending_bound, which calls gs_power_ahr(). There are many bottlenecks, but gs_spending_bound() and hupdate() barely register. The biggest cumulative bottlenecks are expected_time(), expected_event(), gs_info_ahr(), pw_info(), and ahr() (note that these are likely nested, so I'd need to dig deeper to find the best place to optimize first)
devtools::load_all()
library("microbenchmark")
library("profvis")

microbenchmark(
  gs_power_ahr = gs_power_ahr(
    analysis_time = c(12, 24, 36),
    event = c(30, 40, 50),
    binding = TRUE,
    upper = gs_spending_bound,
    upar = list(sf = gsDesign::sfLDOF, total_spend = 0.025, param = NULL, timing = NULL),
    lower = gs_spending_bound,
    lpar = list(sf = gsDesign::sfLDOF, total_spend = 0.025, param = NULL, timing = NULL)
  ),
  times = 50
)
## Unit: milliseconds
## expr     min       lq     mean   median       uq      max neval
## gs_power_ahr 863.136 1004.379 1144.783 1104.985 1238.209 1835.487    50

profvis(
  gs_power_ahr(
    analysis_time = c(12, 24, 36),
    event = c(30, 40, 50),
    binding = TRUE,
    upper = gs_spending_bound,
    upar = list(sf = gsDesign::sfLDOF, total_spend = 0.025, param = NULL, timing = NULL),
    lower = gs_spending_bound,
    lpar = list(sf = gsDesign::sfLDOF, total_spend = 0.025, param = NULL, timing = NULL)
  )
)

profvis({
  for (i in 1:10)
  gs_power_ahr(
    analysis_time = c(12, 24, 36),
    event = c(30, 40, 50),
    binding = TRUE,
    upper = gs_spending_bound,
    upar = list(sf = gsDesign::sfLDOF, total_spend = 0.025, param = NULL, timing = NULL),
    lower = gs_spending_bound,
    lpar = list(sf = gsDesign::sfLDOF, total_spend = 0.025, param = NULL, timing = NULL)
  )
})
Second, I focused in on gs_spending_bound() by using the old examples that were removed in #261. This identified only two bottlenecks:
image.png (view on web)
The biggest bottleneck was calling out to gsDesign::sfLDOF()
https://github.com/Merck/gsDesign2/blob/50067c424ab7dca4579ae895edc5de2bdcd9546e/R/gs_spending_bound.R#L134
The second bottleneck, which was only a fraction of the above, was calling out to hupdate(), which has already been converted to Rcpp
https://github.com/Merck/gsDesign2/blob/50067c424ab7dca4579ae895edc5de2bdcd9546e/R/RcppExports.R#L13
old_examples <- function() {
  info <- (1:3) * 10
  info_frac <- info / max(info)
  k <- length(info_frac)
  
  # 1st analysis
  a1 <- gs_spending_bound(
    k = 1, efficacy = FALSE, theta = 0,
    par = list(sf = gsDesign::sfLDOF, total_spend = 0.025, timing = info_frac, param = NULL),
    hgm1 = NULL
  )
  b1 <- gs_spending_bound(
    k = 1, efficacy = TRUE, theta = 0,
    par = list(sf = gsDesign::sfLDOF, total_spend = 0.025, timing = info_frac, param = NULL),
    hgm1 = NULL
  )
  cat("The (lower, upper) boundary at the 1st analysis is (", a1, ", ", b1, ").\n")
  
  # 2nd analysis
  a2 <- gs_spending_bound(
    k = 2, efficacy = FALSE, theta = 0,
    par = list(sf = gsDesign::sfLDOF, total_spend = 0.025, timing = info_frac, param = NULL),
    hgm1 = h1(r = 18, theta = 0, info = info[1], a = a1, b = b1)
  )
  b2 <- gs_spending_bound(
    k = 2, efficacy = TRUE, theta = 0,
    par = list(sf = gsDesign::sfLDOF, total_spend = 0.025, timing = info_frac, param = NULL),
    hgm1 = h1(r = 18, theta = 0, info = info[1], a = a1, b = b1)
  )
  cat("The upper boundary at the 2nd analysis is (", a2, ", ", b2, ").\n")
}

profvis(for (i in 1:10) old_examples())
In summary, at least for the example call to gs_power_ahr(), converting more of gs_spending_bound() to C++ is unlikely to make a noticeable difference.

**Comment by jdblischak** [comment]
> I think you’re figuring out where investment may be most beneficial

After looking at the profiling results more carefully, I plan to start by optimizing `ahr()` and `pw_info()`

**Comment by nanxstats** [comment]
The initial data.table rewrite seems promising to cut the running time substantially, which is a major source of bottlenecks.

As there is not a current goal to have a complete C++ rewrite of these functions (which would be a much bigger effort), let's close this for now as completed as the primary goal is achieved and we have a clear direction to go.

We can open a new one if and when additional optimization is required.

Thanks @jdblischak for the investigations!
